verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,353 @@
+    If there is a connection in the store, this also validates 'para'
+    against the parameters of an existing connection by relying on the
+    state predicate CheckLocalParameters.
+*)
+ValidConnectionParameters(para) ==
+    /\ para.localEnd.connectionID \in ConnectionIDs
+    /\ para.localEnd.clientID \in ClientIDs
+    /\ \/ store.connection = NullConnection
+       \/ /\ store.connection /= NullConnection
+          /\ CheckLocalParameters(para)
+
+
+(* Operator for reversing the connection ends.
+
+    Given a ConnectionParameters record 'para', returns a new set
+    of parameters where the local and remote ends are
+    flipped (i.e., reversed).
+ *)
+FlipConnectionParameters(para) ==
+    [localEnd   |-> para.remoteEnd,
+     remoteEnd  |-> para.localEnd]
+
+
+(* Operator for construcing a connection proof.
+
+    The connection proof is used to demonstrate to another chain that the
+    local store on this chain comprises a connection in a certain state. 
+ *)
+GetConnProof(connState) ==
+    [height |-> store.latestHeight,
+     connectionState |-> connState]
+
+
+(* Operator for construcing a client proof.
+ *)
+GetClientProof ==
+    [height |-> store.client.latestHeight]
+
+
+(* Verification of a connection proof. 
+
+    This is a state predicate returning true if the following holds: the local
+    client on this chain stores the height reported in the proof. Note that this
+    condition should eventually become true, assuming a correct environment, which
+    should periodically update the client on each chain; see actions 'GoodNextEnv'
+    and 'UpdateClient'.
+
+ *)
+VerifyConnProof(cp) ==
+    /\ cp.height \in store.client.consensusStates
+
+
+(* Verification of a client proof.
+
+    This is a state predicate returning true if the following holds: the height
+    reported in the client proof must not exceed the current (latestHeight) of
+    this chain.
+ *)
+VerifyClientProof(cp) ==
+    cp.height <= store.latestHeight
+
+
+(***************************************************************************
+ Connection Handshake Module actions & operators.
+ ***************************************************************************)
+
+
+(* Modifies the local store.
+
+    Replaces the connection in the store with the argument 'newCon'.
+    This action also advances the chain height. 
+ *)
+NewStore(newCon) ==
+    [store EXCEPT !.connection = newCon,
+                  !.latestHeight = @ + 1]
+
+
+(* Handles a "ICS3MsgInit" message 'm'.
+    
+    Primes the store.connection to become initialized with the parameters
+    specified in 'm'. Also creates a reply message, enqueued on the outgoing
+    buffer.
+ *)
+HandleInitMsg(m) ==
+    LET newCon == [parameters |-> m.parameters,
+                   state      |-> "INIT"]
+        sProof == GetConnProof("INIT")
+        cProof == GetClientProof
+        replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
```

Comment:
We need to agree what we want to model with messages. Either we will use exactly the same format as with ICS, or we will use more high level abstraction that would contain data and the corresponding proof. It feels we are with this version somewhere between the two and then it's a bit confusing. I tend to prefer more the former (although it's not very intuitive) as it would also make integration with the relayer spec easier.